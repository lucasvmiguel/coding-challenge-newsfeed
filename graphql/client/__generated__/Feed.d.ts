/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { Fellowship } from "./globalTypes";

// ====================================================
// GraphQL query operation: Feed
// ====================================================

export interface Feed_feed_User_projects {
  __typename: "Project";
  id: number;
  name: string;
  icon_url: string;
}

export interface Feed_feed_User {
  __typename: "User";
  id: number;
  name: string;
  bio: string;
  avatar_url: string;
  fellowship: Fellowship;
  created_ts: string;
  projects: Feed_feed_User_projects[];
}

export interface Feed_feed_Announcement {
  __typename: "Announcement";
  id: number;
  title: string;
  body: string;
  fellowship: Fellowship;
  created_ts: string;
}

export interface Feed_feed_Project_users {
  __typename: "User";
  id: number;
  name: string;
  avatar_url: string;
}

export interface Feed_feed_Project {
  __typename: "Project";
  id: number;
  name: string;
  description: string;
  icon_url: string;
  created_ts: string;
  users: Feed_feed_Project_users[];
}

export type Feed_feed = Feed_feed_User | Feed_feed_Announcement | Feed_feed_Project;

export interface Feed {
  feed: Feed_feed[];
}

export interface FeedVariables {
  user_type: Fellowship;
  limit?: number | null;
  offset?: number | null;
}
