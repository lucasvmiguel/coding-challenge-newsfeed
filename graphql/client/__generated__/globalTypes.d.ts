/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

export enum Fellowship {
  all = "all",
  angels = "angels",
  founders = "founders",
  writers = "writers",
}

//==============================================================
// END Enums and Input Objects
//==============================================================
