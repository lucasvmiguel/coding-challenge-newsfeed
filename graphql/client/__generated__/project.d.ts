/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: project
// ====================================================

export interface project_project_users {
  __typename: "User";
  id: number;
  name: string;
  avatar_url: string;
}

export interface project_project {
  __typename: "Project";
  id: number;
  name: string;
  description: string;
  icon_url: string;
  users: project_project_users[];
}

export interface project {
  project: project_project;
}

export interface projectVariables {
  id: number;
}
