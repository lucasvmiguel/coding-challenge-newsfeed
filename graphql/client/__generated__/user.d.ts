/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { Fellowship } from "./globalTypes";

// ====================================================
// GraphQL query operation: user
// ====================================================

export interface user_user_projects {
  __typename: "Project";
  id: number;
  name: string;
  icon_url: string;
}

export interface user_user {
  __typename: "User";
  id: number;
  name: string;
  bio: string;
  fellowship: Fellowship;
  avatar_url: string;
  projects: user_user_projects[];
}

export interface user {
  user: user_user;
}

export interface userVariables {
  id: number;
}
