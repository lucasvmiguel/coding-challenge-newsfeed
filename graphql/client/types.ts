export enum resolveTypes {
  User = 'User',
  Announcement = 'Announcement',
  Project = 'Project'
}
