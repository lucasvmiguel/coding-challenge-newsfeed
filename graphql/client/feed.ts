import { gql } from '@apollo/client'

export const FEED_QUERY = gql`
  query Feed($user_type: Fellowship!, $limit: Int, $offset: Int) {
    feed(user_type: $user_type, limit: $limit, offset: $offset) {
      __typename
      ... on User {
        id
        name
        bio
        avatar_url
        fellowship
        created_ts
        projects {
          id
          name
          icon_url
        }
      }
      ... on Announcement {
        id
        title
        body
        fellowship
        created_ts
      }
      ... on Project {
        id
        name
        description
        icon_url
        created_ts
        users {
          id
          name
          avatar_url
        }
      }
    }
  }
`
