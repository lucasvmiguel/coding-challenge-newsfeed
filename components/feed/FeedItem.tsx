
import { resolveTypes } from '../../graphql/client/types'

import AnnouncementCard, { Announcement } from '../card/AnnouncementCard'
import UserCard, { User } from '../card/UserCard'
import ProjectCard, { Project } from '../card/ProjectCard'

export type Item = (Announcement | User | Project) & { __typename: string, created_ts: string }

type Props = {
  item: Item;
}

export default function FeedItem({ item }: Props) {
  switch (item.__typename) {
    case resolveTypes.Project:
      return <ProjectCard project={item as Project} />
    case resolveTypes.User:
      return <UserCard user={item as User} />
    case resolveTypes.Announcement:
      return <AnnouncementCard announcement={item as Announcement} />
    default:
      return null
  }
}