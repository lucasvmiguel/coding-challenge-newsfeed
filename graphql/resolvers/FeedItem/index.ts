import { resolveTypes } from '../../../graphql/client/types'
import { AnnouncementRow, ProjectRow, UserRow } from '../../../graphql/db'

export const __resolveType = (obj: UserRow | AnnouncementRow | ProjectRow, context: any, info: any) => { // eslint-disable-line @typescript-eslint/no-unused-vars
  // only User has bio
  if (isValueValid((obj as UserRow).bio)) {
    return resolveTypes.User
  }

  // only Announcement has body
  if (isValueValid((obj as AnnouncementRow).body)) {
    return resolveTypes.Announcement
  }

  // only Project has description
  if (isValueValid((obj as ProjectRow).description)) {
    return resolveTypes.Project
  }

  return null
}

const isValueValid = (value: any) => {
  return value !== null && value !== undefined
}
