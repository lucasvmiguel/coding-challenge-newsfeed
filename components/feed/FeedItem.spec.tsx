import renderer from 'react-test-renderer'

import { resolveTypes } from '../../graphql/client/types'
import { Fellowship } from '../../graphql/client/__generated__/globalTypes'

import FeedItem from './FeedItem'

it('renders FeedItem component with announcement correctly - snapshot', () => {
  const announcement = {
    __typename: resolveTypes.Announcement,
    id: 1,
    title: "title testing",
    body: "body testing",
    created_ts: "01-01-2021"
  }

  const tree = renderer
    .create(<FeedItem item={announcement} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders FeedItem component with user correctly - snapshot', () => {
  const user = {
    __typename: resolveTypes.User,
    id: 1,
    name: "name testing",
    bio: "desc testing",
    avatar_url: "icon.svg",
    fellowship: Fellowship.founders,
    created_ts: "01-01-2021",
    projects: [
      { id: 2, name: "user testing", icon_url: "icon2.svg" }
    ]
  }

  const tree = renderer
    .create(<FeedItem item={user} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders FeedItem component with project correctly - snapshot', () => {
  const project = {
    __typename: resolveTypes.Project,
    id: 1,
    name: "name testing",
    description: "desc testing",
    icon_url: "icon.svg",
    created_ts: "01-01-2021",
    users: [
      { id: 2, name: "user testing", avatar_url: "icon2.svg" }
    ]
  }

  const tree = renderer
    .create(<FeedItem item={project} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders FeedItem component with no card - snapshot', () => {
  const item = {
    __typename: "invalid",
    id: 1,
    title: "title testing",
    body: "body testing",
    created_ts: "01-01-2021"
  }

  const tree = renderer
    .create(<FeedItem item={item} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})