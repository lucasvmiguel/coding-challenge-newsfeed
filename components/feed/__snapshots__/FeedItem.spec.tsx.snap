// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders FeedItem component with announcement correctly - snapshot 1`] = `
<div
  className="sc-bdfBQB eDCjUP"
>
  <div
    className="sc-dlfnuX iOZAXF"
  >
    <div
      className="sc-hKgJUU jkblDv"
    >
      <img
        className="sc-gsTEea iRQMQw"
        src="https://avatars.dicebear.com/api/initials/a.svg"
      />
    </div>
    <div
      className="sc-eCstlR evEecB"
    >
      <h2>
        title testing
      </h2>
      <div
        className="sc-jSgvzq"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>body testing</p>
",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`renders FeedItem component with no card - snapshot 1`] = `null`;

exports[`renders FeedItem component with project correctly - snapshot 1`] = `
<div
  className="sc-bdfBQB eDCjUP"
>
  <div
    className="sc-dlfnuX iOZAXF"
  >
    <div
      className="sc-hKgJUU jkblDv"
    >
      <img
        className="sc-gsTEea iRQMQw"
        src="icon.svg"
      />
    </div>
    <div
      className="sc-eCstlR evEecB"
    >
      <h2>
        name testing
      </h2>
      <div
        className="sc-jSgvzq"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>desc testing</p>
",
          }
        }
      />
      <h3>
        Participants:
      </h3>
      <div
        className="sc-kEjbQP czxUPI"
      >
        <div
          className="sc-iqHYmW fovqHA"
        >
          <img
            className="sc-jrAFXE dcmrNt"
            src="icon2.svg"
          />
        </div>
        <div
          className="sc-crrszt iBTWmQ"
        >
          <a
            href="/users/2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            user testing
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders FeedItem component with user correctly - snapshot 1`] = `
<div
  className="sc-bdfBQB eDCjUP"
>
  <div
    className="sc-dlfnuX iOZAXF"
  >
    <div
      className="sc-hKgJUU jkblDv"
    >
      <img
        className="sc-gKseQn hvAGSQ"
        src="icon.svg"
      />
    </div>
    <div
      className="sc-eCstlR evEecB"
    >
      <h2>
        name testing
      </h2>
      <p>
        Fellowship: 
        founders
      </p>
      <div
        className="sc-jSgvzq"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>desc testing</p>
",
          }
        }
      />
      <h3>
        Projects:
      </h3>
      <div
        className="sc-iBPTik ddGxJq"
      >
        <div
          className="sc-fubCzh kGMoIv"
        >
          <img
            className="sc-gsTEea iRQMQw"
            src="icon2.svg"
          />
        </div>
        <div
          className="sc-pGacB dXoMBx"
        >
          <a
            href="/projects/2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            user testing
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
